### PR TITLE
Poor man's Dandelion protocol

### PR DESCRIPTION
**Pull request by aivve**
* Broadcast new transactions to two randomly preselected peers.
* Stem peers are randomly selected periodically from the list of active peers (from outgoing connections).
* A new field, named `stem` is added to transaction broadcast message, to which is assigned a ~~random number, being a length of a stem.~~ bool value denoting it being a stem.
* Each node upon relaying a transaction ~~deduces -1 from the `stem` until it hits the zero,~~ randomly switches to the `fluff` phase and broadcast transaction(s) to all peers.
* Stem transactions hashes are stored in poor man's stempool (in memory cache).
* If incoming transactions notification arrives in fluff mode all transaction hashes that match are deleted from stempool.
* Periodically transactions from stempool that didn't made it to return in fluff mode by that time are broadcast as fluff and deleted from stempool.

### CODE REVIEW

**Comment by foiltally** [comment]
What happens if a transaction is forwarded by an older node not supporting this protocol? Is `stem` length reset to zero or it is forwarder as is? 🤔

**Comment by aivve** [comment]
It should forward intact but it will broadcast to all its peers in fluff mode. It only will make it a bit longer to go down to zero, as only nodes supporting stem will decrement it.

File: src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp [hunk 1]
```diff
@@ -933,9 +995,34 @@ void CryptoNoteProtocolHandler::relay_block(NOTIFY_NEW_BLOCK::request& arg) {
...
-  m_p2p->externalRelayNotifyToAll(NOTIFY_NEW_TRANSACTIONS::ID, buf, nullptr);
+void CryptoNoteProtocolHandler::relay_transactions(NOTIFY_NEW_TRANSACTIONS::request& arg) { 
+  if (arg.stem > 0 && !m_dandelion_stem.empty()) { // Dandelion broadcast
+    std::mt19937 rng = Random::generator();
+    std::uniform_int_distribution<> dis(0, 100);
+    auto coin_flip = dis(rng);
+    if (coin_flip < CryptoNote::parameters::DANDELION_STEM_TX_PROPAGATION_PROBABILITY) { // Stem propagation
+      for (const auto& dandelion_peer : m_dandelion_stem) {
+        arg.stem--;
+        if (dandelion_peer.m_state == CryptoNoteConnectionContext::state_normal || dandelion_peer.m_state == CryptoNoteConnectionContext::state_synchronizing) {
```
Comment (aivve) [inline src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp:1079]: We can rewrite here, if peer is dead, then regenerate stem peers and forward to new stem peers

Comment (Lastick) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp:1079]: Yes, an additional check is needed so as not to break the logic. There is a chance of failure if the node is unavailable.

Comment (aivve) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp:1079]: Not a big deal, it will just fluff tx. We randomly with 10% of chances do this anyways.

Comment (Lastick) [reply src/CryptoNoteProtocol/CryptoNoteProtocolHandler.cpp:1079]: Yep, you're right.

**Pull request merged**